hc_dpki/src/keypair.rs

```diff
@@ -0,0 +1,528 @@
+        let mut sk_buf = SecBuf::with_insecure(32);
+        let mut ek_buf = SecBuf::with_insecure(32);
+        util::convert_array_to_secbuf(&sk, &mut sk_buf);
+        util::convert_array_to_secbuf(&ek, &mut ek_buf);
+
+        let pw_sign_pub: bundle::ReturnBundleData = util::pw_enc(&mut sk_buf, &mut passphrase);
+        let pw_enc_pub: bundle::ReturnBundleData = util::pw_enc(&mut ek_buf, &mut passphrase);
+
+        let pw_sign_priv: bundle::ReturnBundleData =
+            util::pw_enc(&mut self.sign_priv, &mut passphrase);
+        let pw_enc_priv: bundle::ReturnBundleData =
+            util::pw_enc(&mut self.enc_priv, &mut passphrase);
+
+        // convert to string
+        let bundle_data = bundle::Keys {
+            pw_sign_pub,
+            pw_enc_pub,
+            pw_sign_priv,
+            pw_enc_priv,
+        };
+        let bundle_data_serialized = json::encode(&bundle_data).unwrap();
+
+        // conver to base64
+        let bundle_data_encoded = base64::encode(&bundle_data_serialized);
+
+        Ok(bundle::KeyBundle {
+            bundle_type,
+            hint,
+            data: bundle_data_encoded,
+        })
+    }
+
+    /// initialize the pairs from an encrypted persistence bundle
+    ///
+    /// @param {object} bundle - persistence info
+    ///
+    /// @param {SecBuf} passphrase - decryption passphrase
+    pub fn from_bundle(
+        bundle: &bundle::KeyBundle,
+        passphrase: &mut SecBuf,
+    ) -> Result<Keypair, HolochainError> {
+        // decoding the bundle.data of type utinl::Keys
+        let bundle_decoded = base64::decode(&bundle.data).unwrap();
+        let bundle_string = str::from_utf8(&bundle_decoded).unwrap();
+        let data: bundle::Keys = json::decode(&bundle_string).unwrap();
+
+        let sk: &bundle::ReturnBundleData = &data.pw_sign_pub;
+        let ek: &bundle::ReturnBundleData = &data.pw_enc_pub;
+        let epk: &bundle::ReturnBundleData = &data.pw_enc_priv;
+        let spk: &bundle::ReturnBundleData = &data.pw_sign_priv;
+        let mut sign_public_key = util::pw_dec(sk, passphrase);
+        let mut enc_public_key = util::pw_dec(ek, passphrase);
+        let enc_priv = util::pw_dec(epk, passphrase);
+        let sign_priv = util::pw_dec(spk, passphrase);
+        Ok(Keypair {
+            pub_keys: util::encode_id(&mut sign_public_key, &mut enc_public_key),
+            enc_priv,
+            sign_priv,
+        })
+    }
+
+    /// sign some arbitrary data with the signing private key
+    ///
+    /// @param {SecBuf} data - the data to sign
+    ///
+    /// @param {SecBuf} signature - Empty Buf the sign
+    pub fn sign(
+        &mut self,
+        data: &mut SecBuf,
+        signature: &mut SecBuf,
+    ) -> Result<(), HolochainError> {
+        let mut data = data;
+        let mut signature = signature;
+        let mut sign_priv = &mut self.sign_priv;
+        sign::sign(&mut data, &mut sign_priv, &mut signature).unwrap();
```

Comment:
I don't understand the shadowing `let mut data = data;`

since the inputs are already mutable references, can you not just directly `sign::sign(data, self.sign_priv, signature)`?

Reply:
I am not very sure about that. But that's the only way I can pass the mut reference to another function.

Reply:
```rust
    pub fn sign(
        &mut self,
        data: &mut SecBuf,
        signature: &mut SecBuf,
    ) -> Result<(), HolochainError> {
        sign::sign(data, &mut self.sign_priv, signature)?;
        Ok(())
    }
```

This works for me locally. Since data and signature are already mutable references, we don't want to get a reference to the reference, so we need to leave off the `&mut`. We still need it on self.sign_priv since, once we deference self, sign_priv points to the actual memory, so we need to get a reference to it.